PEP 753: Uniform URLs in core metadata

woodruffw opened the pull request:
Initial discussion: https://discuss.python.org/t/core-metadata-should-home-page-and-download-url-be-deprecated/62037/16

Signed-off-by: William Woodruff <william@yossarian.net>

<!-- readthedocs-preview pep-previews start -->
----
📚 Documentation preview 📚: https://pep-previews--3936.org.readthedocs.build/pep-0753/

<!-- readthedocs-preview pep-previews end -->

woodruffw committed:
PEP: Uniform URLs in core metadata

Signed-off-by: William Woodruff <william@yossarian.net>

woodruffw committed:
PEP: shan't SHALL

Signed-off-by: William Woodruff <william@yossarian.net>

woodruffw committed:
PEP: add delegate

Signed-off-by: William Woodruff <william@yossarian.net>

woodruffw committed:
PEP: clarify that the well-known labels are a living list

Signed-off-by: William Woodruff <william@yossarian.net>

woodruffw committed:
PEP: Barry is the sponsor

Signed-off-by: William Woodruff <william@yossarian.net>

woodruffw committed:
Apply suggestions from code review

Co-authored-by: Alyssa Coghlan <ncoghlan@gmail.com>

woodruffw committed:
Update peps/pep-9999.rst

Co-authored-by: Alyssa Coghlan <ncoghlan@gmail.com>

woodruffw committed:
PEP: fix ref

Signed-off-by: William Woodruff <william@yossarian.net>

woodruffw committed:
PEP: another canonicalization example

Signed-off-by: William Woodruff <william@yossarian.net>

woodruffw committed:
PEP: formatting fix

Signed-off-by: William Woodruff <william@yossarian.net>

woodruffw committed:
PEP: tweak table widths

Signed-off-by: William Woodruff <william@yossarian.net>

JelleZijlstra commented:
Please take PEP 753.

woodruffw committed:
PEP: assign as PEP 753

Signed-off-by: William Woodruff <william@yossarian.net>

JelleZijlstra commented on peps/pep-9999.rst:
```diff
@@ -0,0 +1,322 @@
+as explicitly document ``Home-page`` and ``Download-URL`` as deprecated in
+favor of equivalent ``Project-URL`` representations.
+
+Deprecating ``Home-page`` and ``Download-URL``
```
For clarity, consider changing this into a Specification section.

JelleZijlstra commented on peps/pep-9999.rst:
```diff
@@ -0,0 +1,322 @@
+  choose to reject the distribution outright. However, this is
+  **NOT RECOMMENDED** until a future unspecified major metadata version
+  formally removes support for ``Home-page`` and ``Download-URL``.
+
```
Redundant newline
```suggestion
```

JelleZijlstra commented on peps/pep-9999.rst:
```diff
@@ -0,0 +1,322 @@
+.. code-block:: python
+
+    import string
+    def canonicalize_label(label: str) -> str:
```
Consider also using Unicode normalization, such as NFC.

woodruffw replied:
No objection, although this has no impact on the current all-ASCII well-known labels -- any concerns @ncoghlan?

woodruffw committed:
PEP 753: change header, remove extra line

Signed-off-by: William Woodruff <william@yossarian.net>

hugovk commented on peps/pep-0753.rst:
```diff
@@ -0,0 +1,321 @@
+        Facundo Tuesca <facundo.tuesca@trailofbits.com>
+Sponsor: Barry Warsaw <barry at python.org>
+PEP-Delegate: Paul Moore <p.f.moore at gmail.com>
+Discussions-To: https://discuss.python.org/t/core-metadata-should-home-page-and-download-url-be-deprecated/62037
```
This should be the new thread created after merge.

hugovk commented on peps/pep-0753.rst:
```diff
@@ -0,0 +1,321 @@
+  ``Project-URL`` fields as a source of URLs over ``Home-page`` and
+  ``Download-URL``, even if the latter are explicitly provided.
+
+* If a distributions's metadata contains only the ``Home-page`` and
```
```suggestion
* If a distribution's metadata contains only the ``Home-page`` and
```

hugovk reviewed (commented):
Gentle reminder for next time, that the [PEP discussion threads](https://discuss.python.org/t/pep-753-uniform-urls-in-core-metadata/62037/17) should begin after merging:

> As soon as a PEP number has been assigned and the draft PEP is committed to the PEP repository, a discussion thread for the PEP should be created to provide a central place to discuss and review its contents, and the PEP should be updated so that the Discussions-To header links to it.

https://peps.python.org/pep-0001/#discussing-a-pep

A good reason is people can link to the published PEP instead of the RTD preview.

For example, I've a comment to make about the well-known labels. This should really go in the discussion thread for all to see, but right now I would have to link to https://pep-previews--3936.org.readthedocs.build/pep-0753/#well-known-labels which will eventually be deleted and be of no use for future readers.

Now I'm tempted to comment instead on this PR, but then it won't be seen by as many interested folk. (I'll wait until after merge.)

But otherwise, thank you very much for writing up this PEP!

woodruffw commented:
> Gentle reminder for next time, that the [PEP discussion threads](https://discuss.python.org/t/pep-753-uniform-urls-in-core-metadata/62037/17) should begin after merging:
> 
> >

Whoops, my bad! I attempted to re-use the old thread, but I can see now that that was the wrong thing to do. I'll create a new thread right now, and update this PR.

woodruffw committed:
PEP 753: fix discussions link

Signed-off-by: William Woodruff <william@yossarian.net>

woodruffw committed:
Update peps/pep-0753.rst

Co-authored-by: Hugo van Kemenade <1324225+hugovk@users.noreply.github.com>

warsaw commented on peps/pep-0753.rst:
```diff
@@ -0,0 +1,322 @@
+    "``bug``", "The project's bug tracker", "``issue``, ``tracker``, ``report``"
+    "``funding``", "Sponsoring information", "``sponsor``, ``donate``, ``donation``"
+
+Packagers and metadata producers **MAY** choose to use these well-known
```
I'm not sure if this recommendation is the same but I would suggest adding:

> Packagers and metadata producers **SHOULD** produce the canonicalized version of the well-known labels in package metadata.

warsaw commented on peps/pep-0753.rst:
```diff
@@ -0,0 +1,322 @@
+  next major core metadata version. If removed, package indices and consumers
+  **MUST** reject metadata containing these fields when said metadata is of
+  the new major version.
+* Enforcement of label canonicalization. If enforced, package producers
```
Essentially this turns my above suggested **SHOULD** to **MUST**.

warsaw commented on peps/pep-0753.rst:
```diff
@@ -0,0 +1,322 @@
+maintainers should observe rejected package uploads or other breaking
+changes to packaging workflows due to this PEP's proposed changes.
+
+Maintainers who observe warnings or changes to the presentation of their
```
I wouldn't limit the learning link to just to "[m]aintainers who observe warnings or changes".  Everyone can learn how this should work from the packaging guide! 😄

warsaw reviewed (approved):
Well done!  I have a small number of suggestions.

woodruffw committed:
PEP 753: relax MUST -> SHOULD

Per https://discuss.python.org/t/pep-753-uniform-urls-in-core-metadata/62792/2

Signed-off-by: William Woodruff <william@yossarian.net>

woodruffw commented on peps/pep-0753.rst:
```diff
@@ -0,0 +1,322 @@
+PEP: 753
+Title: Uniform URLs in core metadata
```
Sounds good, thanks!

woodruffw committed:
PEP 753: rename

Signed-off-by: William Woodruff <william@yossarian.net>

woodruffw committed:
PEP 753: SHOULD -> MUST

Signed-off-by: William Woodruff <william@yossarian.net>

woodruffw committed:
PEP 753: homepage -> home page for informal spelling

Signed-off-by: William Woodruff <william@yossarian.net>

woodruffw commented on peps/pep-0753.rst:
```diff
@@ -0,0 +1,322 @@
+    "``download``", "A download URL for the current distribution, equivalent to ``Download-URL``", "*(none)*"
+    "``changelog``", "The project's changelog", "``changes``, ``releasenotes``, ``whatsnew``, ``history``"
+    "``documentation``", "The project's online documentation", "``docs``"
+    "``bug``", "The project's bug tracker", "``issue``, ``tracker``, ``report``"
```
Thanks for catching this, agreed! I'll make `issues` the canonical label and allow the others as aliases.

woodruffw committed:
PEP 753: tweak canonical labels

Signed-off-by: William Woodruff <william@yossarian.net>

woodruffw committed:
PEP 753: everybody!

Signed-off-by: William Woodruff <william@yossarian.net>

woodruffw committed:
PEP 753: producers should produce the canonicalized labels

Signed-off-by: William Woodruff <william@yossarian.net>

woodruffw commented on peps/pep-0753.rst:
```diff
@@ -0,0 +1,325 @@
+    "``Change_Log``", "``changelog``"
+    "``What's New?``", "``whatsnew``"
+
+Metadata producers **SHOULD** emit the canonicalized form of a user
+specified label, but **MAY** choose to emit the un-canonicalized form so
+long as it adheres to the existing 32 character constraint.
+
+Package indices **SHOULD NOT** use the canonicalized labels directly as UI
+elements (instead replacing them with appropriately capitalized text labels).
```
True, good catch! I'll update the language of this section tomorrow.

woodruffw committed:
PEP 753: clarify use of canonicalized labels

Signed-off-by: William Woodruff <william@yossarian.net>

woodruffw commented:
As a procedural question: does this get merged initially, or should discussion continue on this initial PR? I've had some conflicting experiences here with different PEPs -- some have waited until the initial merge to offer feedback, while others have performed feedback before the draft becomes visible on `peps.python.org`.

JelleZijlstra commented:
We're sort of inconsistent about that; personally I'd prefer to merge PEPs sooner so they can become visible on peps.python.org. You can then always make future changes as a result of the discussion.

Since Barry approved and the only open discussion is a non-blocking point from me, I'll merge this PR now.

encukou commented:
AFAIK, the general process is:
- the *presentation* gets fixed up here (typos, duplicate sections, but also unfortunate wording or internal inconsistencies)
- the *ideas* get discussed on Discourse

Of course, the line isn't clear; people naturally want to point things out as soon as they see them; and some fixes are straightforward enough to merge soon :)

For things like Unicode normalization here, it's perfectly OK to ask PEP editors to merge *this* version of the PEP, and then you or Jelle can raise the point in the discussion thread.
To me, [your comment](https://github.com/python/peps/pull/3936#discussion_r1742290576) made it look like you want to resolve this before merging, possibly to save pointless public discussion.

hugovk commented:
Yep, it's defined in PEP 1 at [Submitting a PEP](https://peps.python.org/pep-0001/#submitting-a-pep) and [PEP Editor Responsibilities & Workflow](https://peps.python.org/pep-0001/#pep-editor-responsibilities-workflow). It includes making sure the PEP makes some sort of technical sense, but hammering down the ideas is done on Discourse with a bigger audience.

This is partly what I was trying to get at in https://github.com/python/peps/pull/3936#pullrequestreview-2278152702: get the PEP merged fairly quickly, then we have a peps.python.org URL to _begin_ the PEP discussion on Discourse. That way we don't begin a potentially long thread with a top post containing outdated draft RTD links that can no longer be edited and become 404s.

But as said, we're not always consistent in this :)
